test/integration/humming_web_app.py

```diff
@@ -0,0 +1,226 @@
+        await ws.send_str(message)
+
+    async def send_ws_json(self, ws_path, data):
+        ws = [ws for path, ws in self._ws_response.items() if ws_path in path][0]
+        await ws.send_json(data=data)
+
+    # To add or update data which will later be respoonded to a request according to its method, host and path
+    def update_response(self, method, host, path, data, params=None, is_json=True):
+        method = method.upper()
+        resp_data = [x for x in self._stock_responses if x.method == method and x.host == host and x.path == path
+                     and x.params == params]
+        if resp_data:
+            self._stock_responses.remove(resp_data[0])
+        self._stock_responses.append(StockResponse(method, host, path, params, is_json, data))
+
+    def add_host_to_mock(self, host, ignored_paths=[]):
+        HummingWebApp._hosts_to_mock[host] = ignored_paths
+
+    # reroute a url if it is one of the hosts we handle.
+    @staticmethod
+    def reroute_local(url):
+        a_url = URL(url)
+        if a_url.host in HummingWebApp._hosts_to_mock and not any(x in a_url.path for x in
+                                                                  HummingWebApp._hosts_to_mock[a_url.host]):
+            host_path = f"/{a_url.host}{a_url.path}"
+            query = a_url.query
+            a_url = a_url.with_scheme("http").with_host(HummingWebApp.host).with_port(HummingWebApp._port)\
+                .with_path(host_path).with_query(query)
+        return a_url
+
+    orig_session_request = requests.Session.request
+
+    # self here is not an instance of HummingWebApp, it is for mocking Session.request
+    @staticmethod
+    def reroute_request(self, method, url, **kwargs):
+        a_url = HummingWebApp.reroute_local(url)
+        return HummingWebApp.orig_session_request(self, method, str(a_url), **kwargs)
+
+    @property
+    def started(self) -> bool:
+        return self._started
+
+    @property
+    def port(self) -> Optional[int]:
+        return type(self)._port
+
+    async def _start(self):
+        try:
+            HummingWebApp._port = get_open_port()
+            self._impl: Optional[web.Application] = web.Application()
+            self._impl.add_routes([web.route("*", '/{tail:.*}', self._handler)])
+            self._runner = web.AppRunner(self._impl)
+            await self._runner.setup()
+            site = web.TCPSite(self._runner, host=HummingWebApp.host, port=HummingWebApp._port)
+            await site.start()
+            self._started = True
+        except Exception:
+            logging.error("oops!", exc_info=True)
+
+    async def wait_til_started(self):
+        while not self._started:
+            await asyncio.sleep(0.1)
+
+    async def _stop(self):
+        if self._runner is None:
+            return
+        await self._runner.cleanup()
+        self._runner = None
+        self._impl = None
+        self._port = None
+        self._started = False
```

Comment:
You should also call `self._ev_loop.stop()` here, to stop the event loop and end the wrapping thread.

Reply:
Done.